### ISSUE TITLE
[r1.7][TensorRT] Questions about the calibration in INT8 mode of TensorRT optimization

### ISSUE THREAD

**Issue by oscarriddle** [tensorflow/tensorflow#19141]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:  NO
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Red Hat 4.8.5-16, Linux version 3.10.0-693.5.2.el7.x86_64
- **TensorFlow installed from (source or binary)**: pip (python 2.7)
- **TensorFlow version (use command below)**: tensorflow-gpu==1.7.0
- **Python version**:  python 2.7
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: gcc 5.3
- **CUDA/cuDNN version**: CUDA9.0, cuDNN7.0.5
- **GPU model and memory**: Tesla P4, 8GB
- **Exact command to reproduce**: NA

### Describe the problem
I tried to evaluate the accuracy of TF integrated TRT INT8 optimization in Python. However, I followed the procedure as the example test: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/tensorrt/test/test_tftrt.py
but the inference accuracy is not good (FP32: 70%, INT8-without-calib: 16%, INT8-with-calib: 35%).

I tried to trace the code but on the API level, but I'm a little confused about how does the calibration work. looks like the function run_calibration() just created a session and accepted the calib dataset to run the session, and then the return value was abandoned. 

I checked this introduction:[](https://devblogs.nvidia.com/int8-inference-autonomous-vehicles-tensorrt), seems that the TRT python INT8 calibration will generate a calibration cache in file and then be used in the later inference. But in this TFTRT, the calibration didn't generate a calibration cache.

Another observation is that, I previously tried the TensorRT optimization in C++ independently, which is not integrated in the Tensorflow. At that time, the accuracy lose on INT8 mode optimization is about 10% (FP32: 70%, INT8: 60%). Note that my C++ and Python experiments are using the same model and same calibration dataset and the same test dataset, so from my point of view, perhaps TRT INT8 Python API (at least the TF1.7 integrated version) seems somehow different from the TRT original C++ INT8 API. 

So let me conclude my questions:
1) Is the TF integrated TRT INT8 optimization in Python has a different back-end process compared from the independent TRT INT8 optimization in C++?
2) Could you share a hint of how does the python version INT8 calibration is realized? Is the generated calibration data been stored in the cache so that can be loaded in later inference?

### Source code / logs
Please refer to the test:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/tensorrt/test/test_tftrt.py

**Comment by samikama** [comment]
@oscarriddle , 
You can't have INT8 without calibration. It should fail. Are you sure that you are running an int8 graph? 

Also TFTRT int8 is a bit more complicated than TRT only int8. In native TRT(C++ version you mention) you are calibrating a graph end-to-end and int8 calibration TRT tries the minimize the information loss between input and output but still introduces some small errors as you describe. In TFTRT you are likely converting a subgraph rather than full graph, meaning there are TF native nodes in the graph after TRT graph. Since the calibration can only minimize the information loss in the engine, the errors introduced at the output of the engine due to calibration, may, and most likely will be, amplified by the downstream nodes. If you want to minimize this, you will need to do calibration aware training. We are thinking on how can we simplify that and have some ideas to test. 

As for the answers,
1) No, they have the same calibration scheme. Likely the calibrated parts of the graph are different. Did you check how much of the graph is converted to TRT engine?
2) In TFTRT, calibration data is not serialized. It is consumed immediately. When you create a calibration graph, a calibrator is created and attached to the graph. When you run the calibration graph data is passed to the calibrator to collect calibration data and when you convert calibration graph to inference graph, nodes in calibration graph is taken out and replaced with TRT engine. If you save inference graph, it will always use the calibration you made and it doesn't require further calibration.

Cheers,
Sami

**Comment by oscarriddle** [comment]
Hi @samikama ,
Thanks for your detailed reply, it indeed help clarified the answer to my questions.

> No, they have the same calibration scheme. Likely the calibrated parts of the graph are different. Did you check how much of the graph is converted to TRT engine?

My model is InceptionV3 that contains about 800 nodes, and after the TFTRT optimization, nodes number shrunk to about 7~10 (varies if I set an internal nodes to be the output node).

Just for now, I think the TFTRT INT8  seems still have some unaddressed issues that would cause a big sacrifice on the accuracy (Not sure whether because a higher efficient models have a higher information entropy per bit). 

Another plan is to just use the FP32 mode TFTRT optimization, which has no accuracy sacrifice but with 50~70% throughput boost. However, I'm wondering whether TFTRT could optimize a graph partially. For example, if I set the parameter "outputs" in API "create_inference_graph()" to be an internal node, could the downstream nodes just be reserved? 

I think this would be practically useful. For instance, a InceptionV3 graph with a final part like "GlobalPool(image feature in high dim)->dropout->Conv2d(FC Layer)->SpatialSqeeze(Classification result)->Softmax". Not only the classification result, but also the GlobalPool layer result is useful too, since it represents the feature of a image in higher dimensions. 

I've just give a try to get both the GlobalPool and SpatialSqeeze results in one turn. My procedure is to extract the weights of Conv2d layer from original frozen model, then optimize the original graph from start to the GlobalPool, then use numpy methods to do the dropout, Conv2d, SpatialSqeeze calculation. 

It's just a raw idea that I would like to share with you. 

Thanks,

**Comment by samikama** [comment]
@oscarriddle,

As I mentioned above, int8 engines in TF-TRT falls into partial quantization domain and requires you to do quantization aware training of your graph. It is a complicated process today and we are trying to find ways to simplify this for the users. However for advanced users or for the cases where entire graph replaced with a single TRTEngineOP, current quantization should not have too much accuracy loss, provided you can provide a diverse enough calibration set.

**Comment by wincle** [comment]
I met the same problem,  the results of INT8 model seems to be wired.
